File: typescript/examples/url-join-path.test.ts
// カード url-join-path の Contract を検証するテスト
import { describe, expect, it } from 'vitest';

const href = (input: string, base?: string) => new URL(input, base).href;

describe('url-join-path: URL', () => {
  it('ベースの最後のセグメントが置き換わり、末尾 / の有無で結果が変わる', () => {
    expect(href('users', 'https://api.example.com/v1/')).toBe('https://api.example.com/v1/users');
    expect(href('users', 'https://api.example.com/v1')).toBe('https://api.example.com/users');
    expect(href('users', 'https://api.example.com')).toBe('https://api.example.com/users');
    expect(href('./users', 'https://api.example.com/v1/')).toBe('https://api.example.com/v1/users');
    expect(href('users/', 'https://api.example.com/v1/')).toBe('https://api.example.com/v1/users/');
  });

  it('先頭 / はホスト直下、//host はスキームだけ引き継ぐ、スキーム付きはベースを無視', () => {
    expect(href('/health', 'https://api.example.com/v1/')).toBe('https://api.example.com/health');
    expect(href('//cdn.example.com/x.js', 'https://api.example.com/v1/')).toBe('https://cdn.example.com/x.js');
    expect(href('//cdn.example.com/x.js', 'http://api.example.com/')).toBe('http://cdn.example.com/x.js');
    expect(href('https://other.example.com/z', 'https://api.example.com/v1/')).toBe('https://other.example.com/z');
  });

  it('.. は 1 つ上へ戻り、ルートより上には戻れない', () => {
    expect(href('../v2/users', 'https://api.example.com/v1/')).toBe('https://api.example.com/v2/users');
    expect(href('../../../x', 'https://api.example.com/a/b/')).toBe('https://api.example.com/x');
  });

  it('クエリとフラグメントの引き継ぎ', () => {
    const base = 'https://api.example.com/v1/items?page=2#top';
    expect(href('users', base)).toBe('https://api.example.com/v1/users');
    expect(href('users?x=1', base)).toBe('https://api.example.com/v1/users?x=1');
    expect(href('?x=1', base)).toBe('https://api.example.com/v1/items?x=1');
    expect(href('#frag', base)).toBe('https://api.example.com/v1/items?page=2#frag');
    expect(href('', base)).toBe('https://api.example.com/v1/items?page=2');
  });

  it('空白や非 ASCII はパーセントエンコードされる', () => {
    expect(href('c d/é', 'https://api.example.com/')).toBe('https://api.example.com/c%20d/%C3%A9');
  });

  it('ベース無しの相対参照や不正なベースは TypeError', () => {
    expect(() => new URL('users')).toThrowError(TypeError);
    expect(() => new URL('users', 'not a url')).toThrowError(TypeError);
    expect(() => new URL('users', 'mailto:a@example.com')).toThrowError(TypeError);
    let code: unknown;
    try {
      new URL('users');
    } catch (e) {
      code = (e as { code?: unknown }).code;
    }
    expect(code).toBe('ERR_INVALID_URL');
  });

  it('URL オブジェクトをベースにしても変更されない', () => {
    const base = new URL('https://api.example.com/v1/');
    expect(href('users', base.href)).toBe('https://api.example.com/v1/users');
    expect(new URL('users', base).href).toBe('https://api.example.com/v1/users');
    expect(base.href).toBe('https://api.example.com/v1/');
  });

  it('文字列連結は // の重複や / の欠落を起こす', () => {
    expect('https://api.example.com/v1/' + '/users').toBe('https://api.example.com/v1//users');
    expect('https://api.example.com/v1' + 'users').toBe('https://api.example.com/v1users');
  });
});

File: python/examples/url-join-path_test.py
"""url-join-path: urllib.parse.urljoin の Contract を検証する。"""

from urllib.parse import urljoin, urlsplit

import pytest

BASE = "https://api.example.com/v1/"


def test_resolves_relative_reference() -> None:
    """ベースの最後のセグメントが置き換わる。末尾 / の有無で結果が変わる。"""
    assert urljoin(BASE, "users") == "https://api.example.com/v1/users"
    assert urljoin("https://api.example.com/v1", "users") == "https://api.example.com/users"
    assert urljoin(BASE, "./users") == "https://api.example.com/v1/users"
    assert urljoin(BASE, "users/") == "https://api.example.com/v1/users/"


def test_leading_slash_scheme_relative_and_absolute() -> None:
    """先頭 / はホスト直下、//host はスキームだけ引き継ぐ、スキーム付きはそのまま。"""
    assert urljoin(BASE, "/health") == "https://api.example.com/health"
    assert urljoin(BASE, "//cdn.example.com/x.js") == "https://cdn.example.com/x.js"
    assert urljoin(BASE, "http://other.com/a") == "http://other.com/a"
    assert urljoin(BASE, "HTTP://Other.com/A") == "HTTP://Other.com/A"
    assert urljoin(BASE, "mailto:x@y") == "mailto:x@y"


def test_dotdot_and_double_slash() -> None:
    """.. は 1 つ上へ戻り、ルートより上には戻れない。// は 1 つに畳まれる。"""
    assert urljoin(BASE, "../v2/users") == "https://api.example.com/v2/users"
    assert urljoin(BASE, "..") == "https://api.example.com/"
    assert urljoin("https://x.com/a/b/c", "../../d") == "https://x.com/d"
    assert urljoin("https://x.com/a/b/c", "../../../../d") == "https://x.com/d"
    assert urljoin("https://x.com/a/", "b//c") == "https://x.com/a/b/c"


def test_query_and_fragment_of_base_are_not_inherited() -> None:
    """ベースのクエリ・フラグメントはパスがあれば捨てる。? だけならパスを保ち、# だけならクエリも保つ。"""
    base = "https://x.com/a?q=1#f"
    assert urljoin(base, "b") == "https://x.com/b"
    assert urljoin(base, "?r=2") == "https://x.com/a?r=2"
    assert urljoin(base, "#g") == "https://x.com/a?q=1#g"
    assert urljoin(base, "") == base
    assert urljoin(BASE, "users?page=2") == "https://api.example.com/v1/users?page=2"


def test_does_not_percent_encode() -> None:
    """空白や非 ASCII はエンコードされない。"""
    assert urljoin(BASE, "c d/é") == "https://api.example.com/v1/c d/é"


def test_never_raises_on_odd_inputs() -> None:
    """空のベース・スキーム無しのベース・階層を持たないスキームでも例外を投げない。"""
    assert urljoin("", "users") == "users"
    assert urljoin("v1/", "users") == "v1/users"
    assert urljoin("mailto:a@b", "x") == "x"
    assert urlsplit(urljoin("mailto:a@b", "x")).netloc == ""


def test_scheme_is_lowercased_but_host_is_not() -> None:
    """ベースのスキームは小文字化されるがホストはそのまま。"""
    assert urljoin("HTTPS://X.com/a/", "b") == "https://X.com/a/b"


def test_str_and_bytes_cannot_be_mixed() -> None:
    """str 同士か bytes 同士。混ぜると TypeError。"""
    assert urljoin(b"https://x.com/a/", b"b") == b"https://x.com/a/b"
    with pytest.raises(TypeError):
        urljoin("https://x.com/a/", b"b")  # type: ignore[arg-type]


def test_user_input_can_hijack_host() -> None:
    """Pitfalls: ユーザー入力でスキームやホストを乗っ取れる。結合後に検証する。"""
    for bad in ("javascript:alert(1)", "//evil.com/"):
        joined = urljoin(BASE, bad)
        assert urlsplit(joined).netloc != "api.example.com"
